Fixes for projects API

smurching opened the pull request:
Fixes hanging projects API tests / simplifies the async run framework/fixes a bug where we'd create two runs when running on Databricks

* Local & Databricks runs are no longer asynchronously monitored in a subprocess by the client. Instead, the client assumes that local & Databricks project runs will report their own status to a tracking server.
* Asynchronous local runs are now launched by invoking `mlflow run -m local` (the CLI is always blocking) in a subprocess opened via `subprocess.Popen`, instead of via `multiprocessing.Process` (should reduce the likelihood of resource leaks, since `subprocess.Popen` calls fork() + exec() instead of just fork()).
* Fixes a bug where we'd create a run twice when running on Databricks (once from the client before launching the run, and again when invoking `mlflow run` on the Databricks cluster)

After this PR:
* `SubmittedRun.get_status` now returns the status of the launched unit of execution (local process / Databricks run) rather than an MLflow run status from a tracking server
* A program that calls `mlflow.projects.run()` and exits will leave behind launched runs, which will asynchronously report their status to a tracking server. The old behavior was to block until all launched runs completed.

smurching committed:
WIP; local running seems ok

smurching committed:
Updates

smurching committed:
WIP

smurching committed:
WIP

smurching committed:
WIP

smurching committed:
WIP, going to remove active_run and get_status from SubmittedRun

smurching committed:
Updates

smurching committed:
Update

smurching committed:
Update

smurching committed:
WIP

smurching committed:
revert

smurching committed:
Fix databricks test

smurching committed:
Updates

smurching committed:
Clarify comment

smurching committed:
Fix comment

smurching commented on mlflow/cli.py:
```diff
@@ -180,6 +181,7 @@ def server(file_store, default_artifact_root, host, port, workers):
 cli.add_command(mlflow.sagemaker.cli.commands)
 cli.add_command(mlflow.azureml.cli.commands)
 cli.add_command(mlflow.experiments.commands)
+cli.add_command(mlflow.utils.internal_cli.commands)
```
We use a new CLI command instead of invoking `mlflow run` directly because:
* There's some setup in `mlflow run` that should happen just-once in an environment (e.g. multiple concurrent attempts to create the same conda env can fail)
* I didn't want `mlflow run` to have different behavior when invoked by the user in a shell vs in `mlflow.projects.run()`. Without a separate CLI command, we'd need to have `mlflow run` conditionally reuse an existing run ID / log parameters depending on whether it was being invoked by a user or in `mlflow.projects.run`

mateiz replied:
Could we do this through a flag on `mlflow run` instead, similar to how we have no-conda and stuff like that? If not, is there any way to hide this from the help for `mlflow`? It's really weird to see an internal command there.

mateiz replied:
Also, I didn't quite understand which bits of `mlflow run` we don't want in this case. For creating Conda environments, does this solution somehow prevent people from running `mlflow run` multiple times on their own, or spawning multiple async runs? If not, it seems like we should make that recover from the concurrent Conda issue. For the run ID part, maybe `mlflow run` can take an optional run ID to use, and not start its own when this special flag is passed (as mentioned above).

mateiz replied:
Looks like maybe `hidden=True` can hide commands in Click: https://github.com/pallets/click/pull/500

smurching replied:
Agree that it's weird to see an internal command, thanks for tip RE hiding commands. Regarding your questions:

> Could we do this through a flag on mlflow run instead, similar to how we have no-conda and stuff like that?

We could have `mlflow run`'s behavior vary via a flag (or by adding a new undocumented 'mode' e.g. `mlflow run -m run_entry_point ...`).

It's tricky to have `mlflow run` invoke itself as there's a set `S` of things we want to happen just-once during a single project run. For example, across (a) an invocation of `mlflow run` launched by the user via the CLI and (b) one launched via `subprocess.Popen` in `mlflow.projects.run` we want to create a new run and log parameters just once.

We could get around this by passing different arguments (i.e. a run ID when we don't want to create a run) / using `mode` to determine the right behavior, my concern is if/as `S` grows, we have to keep adding arguments to `mlflow run` (& mark these arguments as internal-use only). By running separate commands we can more easily decouple cases (a) and (b).

However I'm open to running everything via `mlflow run`, don't have a super strong opinion either way.

> Also, I didn't quite understand which bits of mlflow run we don't want in this case. For creating Conda environments, does this solution somehow prevent people from running mlflow run multiple times on their own, or spawning multiple async runs? If not, it seems like we should make that recover from the concurrent Conda issue. 

With this PR users can spawn multiple async runs of the same project via the Python API; the conda environment will be created synchronously for the first run of the project, so parallel conda env creation shouldn't be a problem.

This PR doesn't prevent people from invoking `mlflow run` concurrently e.g. in separate terminal windows (conda env creation would break here), although hopefully this use case isn't needed if users can launch concurrent runs via the Python API. 

> Looks like maybe hidden=True can hide commands in Click: pallets/click#500

Thanks for the tip! Unfortunately seems this functionality will only be released with Click 7.0 / it's not currently available

mateiz replied:
Ah, I see. In that case maybe a special "mode" is the best way. For example `local-noenv` or something.

smurching committed:
Merge master, WIP making some changes

smurching committed:
Updates; merge master, use 'mlflow run' instead of creating new CLI command

smurching committed:
Remove unused internal CLI

smurching committed:
Updates

smurching committed:
Typo + test fix

smurching committed:
Update CLI tests

smurching committed:
Updates

smurching committed:
lint

aarondav commented on mlflow/cli.py:
```diff
@@ -71,8 +71,12 @@ def cli():
                    "MLflow will download artifacts from distributed URIs passed to parameters of "
                    "type 'path' to subdirectories of storage_dir.")
+@click.option("--run-id", metavar="RUN_ID",
+              help="Note: this argument is used internally by the MLflow project APIs "
```
Maybe flip these, first say what it does and then say that this is mainly for internal use?

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -27,7 +29,8 @@
 # "conda" if unset
-MLFLOW_CONDA = "MLFLOW_MLFLOW_CONDA"
+MLFLOW_CONDA = "MLFLOW_CONDA"
+MLFLOW_ENTRY_POINT_MODE = "_local_run_entry_point"
```
Add a comment on what this used for? (e.g., to communicate between two processes for some purpose)

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -251,46 +264,127 @@ def _maybe_create_conda_env(conda_env_path):
+        active_run.set_terminated(status)
+
+
+def _run_local_noenv(
```
Please document what this is used for and how it's different from the other similar methods, like _launch_local_run. It's kind of awkward that we had to split this out -- what is the reason?

aarondav replied:
So all this code in `_run_local_noenv` and `_run_entry_point_command` is specific to the case where we are doing `projects.run`, where the mode is local? Is that true?

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -251,46 +264,127 @@ def _maybe_create_conda_env(conda_env_path):
+
+def _run_entry_point_command(command):
+    """
+    Meant to be run in a subprocess via a CLI command triggered in _run_project. Runs an entry point
```
This isn't meant to be run in a subprocess, it starts a subprocess, right? Also is the tracking server stuff still true?

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -251,46 +264,127 @@ def _maybe_create_conda_env(conda_env_path):
-def _launch_local_run(active_run, command, work_dir, env_map, stream_output):
+def _launch_local_run(
+        uri, run_id, entry_point, parameters, work_dir, env_map, use_conda, storage_dir,
+        stream_output):
```
Add doc on stream_output.

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -251,46 +264,127 @@ def _maybe_create_conda_env(conda_env_path):
-def _run_project(project, entry_point, work_dir, parameters, use_conda, storage_dir,
-                 experiment_id, block):
-    """Locally run a project that has been checked out in `work_dir`."""
+def _maybe_set_run_terminated(active_run, status):
```
I thought this code already existed somewhere else? Was it removed?

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -97,20 +100,29 @@ def run(uri, entry_point="main", version=None, parameters=None, experiment_id=No
+    :return: A `SubmittedRun` exposing information (e.g. run ID) about the launched run. Note that
+             the returned `SubmittedRun` is not thread-safe.
     """
+    if mode == MLFLOW_ENTRY_POINT_MODE:
```
`projects.run` calls `_run` which calls `_run_local` which calls `_run_project` which calls `_launch_local_run` which calls `subprocess.Popen` with a special mode which calls `projects.run` which bypasses the `_run` invocation to call `_run_local_noenv`.

That's a pretty long chain to understand, and most of those methods are only used by the method immediately prior to them. This suggests that we could refactor them into a driver method, and a bunch of helper methods to make the driver method more readable.

smurching committed:
WIP refactoring into separate modules - should prob do this in another PR

smurching committed:
Revert "WIP refactoring into separate modules - should prob do this in another PR"

This reverts commit c8d4470840f80a849c231cf873231ffa1e4afc21.

smurching committed:
WIP

smurching committed:
Merge branch 'master' into fix-projects-api

smurching committed:
Notes from discussion

smurching committed:
WIP

smurching committed:
WIP; remove cancellation logic, trying to simplify further

smurching committed:
Updates

smurching committed:
Fix lint

smurching committed:
WIP adding back excepthook so we properly clean up children

smurching committed:
Updates

smurching committed:
lint

smurching committed:
Remove auto run-cancellation

smurching commented on tests/projects/test_projects.py:
```diff
@@ -26,6 +26,12 @@ def _assert_dirs_equal(expected, actual):
+    return base_uri
+
+
 def test_fetch_project(tmpdir):
```
Refactored the _fetch_project a bit for brevity / since the `_fetch_project` API changed in this PR

smurching commented on tests/projects/test_projects_cli.py:
```diff
@@ -13,15 +13,13 @@ def test_run_local():
-                                              "greeting=hi", "-P", "name=%s" % name,
-                                              "-P", "excitement=%s" % excitement_arg])
-            assert name in res.output
+            invoke_cli_runner(cli.run, [TEST_PROJECT_DIR, "-e", "greeter", "-P",
```
Changed since the entry point command is now printed out by a subprocess

smurching committed:
Updates from self-review pass

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -27,7 +27,7 @@
 # Environment variable indicating a path to a conda installation. MLflow will default to running
 # "conda" if unset
-MLFLOW_CONDA = "MLFLOW_MLFLOW_CONDA"
+MLFLOW_CONDA = "MLFLOW_CONDA"
```
Is this intended? This seems like a backwards-incompatible change, if it's in effect, and probably deserves a PR of its own (however small).

smurching replied:
Yeah it's intended, looks like I made a typo in an earlier PR - we haven't released an MLflow version with this change yet so thankfully the typo hasn't messed anything up. That being said, good call on putting it in a separate PR - I'll do that

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -37,28 +37,51 @@ class ExecutionException(Exception):
             experiment_id=exp_id, cluster_spec=cluster_spec, git_username=git_username,
             git_password=git_password)
+    elif mode == "local" or mode is None:
+        # Fetch project into a working directory
```
Remove comment, it is just restating the next line

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -37,28 +37,51 @@ class ExecutionException(Exception):
+        # Fetch project into a working directory
+        work_dir = _fetch_project(uri, use_temp_cwd, version, git_username, git_password)
+        project = _load_project(project_dir=work_dir)
+        # Validate we specified correct params for the project
```
Remove comment, it is just restating the next line

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -37,28 +37,51 @@ class ExecutionException(Exception):
+        project = _load_project(project_dir=work_dir)
+        # Validate we specified correct params for the project
+        project.get_entry_point(entry_point)._validate_parameters(parameters)
+        if use_conda:
```
This one might be worth a comment to mention that we have to load the conda env in the blocking section in order to avoid concurrency problems.

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -279,58 +300,102 @@ def _maybe_create_conda_env(conda_env_path):
+    return active_run
+
+
+def _run_local(
```
Maybe rename this to something more descriptive like `_invoke_mlflow_run_subprocess`

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -279,58 +300,102 @@ def _maybe_create_conda_env(conda_env_path):
+    """
+    Run an MLflow project from the given URI in a new directory.
+
+    Supports downloading projects from Git URIs with a specified version, or copying them from
```
This part of the description is no longer relevant, right? Project download is not done in this method.

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -279,58 +300,102 @@ def _maybe_create_conda_env(conda_env_path):
-
-    return _launch_local_run(
-        active_run, command, work_dir, env_map, stream_output=block)
+    # Invoke `mlflow run` with a special mode, which will in turn run the entry point command in
```
Can remove the special mode thing, and just mention that this will always run in blocking mode.

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -279,58 +300,102 @@ def _maybe_create_conda_env(conda_env_path):
+    final_env.update(env_map)
+    # TODO: maybe write the output of asynchronous local runs e.g. as artifacts for
+    # debugging, we currently just drop it
+    # Launch `mlflow run` command as the leader of its own process group so that we can do a
```
"Launch `mlflow run` command as the leader of its own process group" -- this is not how I would describe popening mlflow run, is this is a comment from a previous incarnation?

smurching replied:
We currently actually do launch `mlflow run` such that the subprocess invoking `mlflow run` is the leader of its own process group (so that we can terminate the entire process tree rooted at the `mlflow run` command when `SubmittedRun.cancel()` is called). This is useful for cleaning up child processes in cancel() e.g. if the `mlflow run` command invokes an entry point command that itself creates subprocesses. Can discuss offline too, maybe we don't need to handle this case.

aarondav replied:
OK

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -279,58 +300,102 @@ def _maybe_create_conda_env(conda_env_path):
+    # best-effort cleanup of all its descendant processes if needed
+    return subprocess.Popen(
+        mlflow_run_arr, env=final_env, universal_newlines=True,
+        stderr=open(os.devnull, "w"), stdout=open(os.devnull, "w"), preexec_fn=os.setsid)
```
Was the previous behavior to interleave output, or drop the output? I kind of prefer the overly verbose strategy until we have a better story, otherwise it may be very hard to understand what goes wrong when things go wrong.

smurching replied:
Previously we dropped the output, I'm fine with having the output interleave

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -279,58 +300,102 @@ def _maybe_create_conda_env(conda_env_path):
+def _run_entry_point(command, work_dir, active_run):
+    """
+    Runs an entry point command in a subprocess, waits on the subprocess to finish, & reports its
+    status to the tracking server.
```
Update comment

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -279,58 +300,102 @@ def _maybe_create_conda_env(conda_env_path):
+    return LocalSubmittedRun(run_id, process)
+
+
+def _build_mlflow_run_cmd(uri, entry_point, storage_dir, use_conda, run_id, parameters):
```
Add comment

aarondav commented on mlflow/projects/databricks.py:
```diff
@@ -58,6 +58,11 @@ def _get_databricks_run_cmd(dbfs_fuse_tar_uri, entry_point, parameters):
     project_dir = os.path.join(DB_PROJECTS_BASE, tar_hash)
     mlflow_run_arr = list(map(shlex_quote, ["mlflow", "run", project_dir, "--new-dir",
                                             "--entry-point", entry_point]))
+    # TODO: we need this check since run ID may be undefined e.g. if we launch a Databricks run
```
It seems safe to always condition extension on existence, I don't know if we need all this TODO. We can just keep the if statement.

aarondav commented:
The flow looks good! Mostly minor comments.

smurching committed:
Respond to review comments:
* Update outdated comments / rename a few methods
* Remove rename of MLFLOW_CONDA constant, to be reintroduced in a separate PR

Also: if parameters are unspecified, set them to an empty dict early on so that we don't have to handle None values everywhere

smurching committed:
Merge master, fix tests

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -43,36 +43,35 @@ def _run(uri, entry_point="main", version=None, parameters=None, experiment_id=N
     Returns a `SubmittedRun` corresponding to the project run.
     """
     exp_id = experiment_id or tracking._get_experiment_id()
+    params = parameters or {}
```
Maybe we can overwrite `parameters` -- otherwise it's pretty likely that we will accidentally use `parameters` as opposed to `params` later on in the function.

smurching committed:
Review comments
